mesonbuild/mdist.py

```diff
@@ -189,72 +239,58 @@ def process_submodules(self, src: str, distdir: str) -> None:
             elif status in {'+', 'U'}:
                 handle_dirty_opt(f'Submodule {subpath!r} has uncommitted changes that will not be included in the dist tarball', self.options.allow_dirty)
 
-            self.copy_git(os.path.join(src, subpath), distdir, revision=sha1, prefix=subpath)
+            self.copy_repo(os.path.join(src, subpath), distdir, revision=sha1, prefix=subpath)
 
-    def create_dist(self, archives: T.List[str]) -> T.List[str]:
-        self.process_git_project(self.src_root, self.distdir)
-        for path in self.subprojects.values():
-            sub_src_root = os.path.join(self.src_root, path)
-            sub_distdir = os.path.join(self.distdir, path)
-            if os.path.exists(sub_distdir):
-                continue
-            if is_git(sub_src_root):
-                self.process_git_project(sub_src_root, sub_distdir)
-            else:
-                shutil.copytree(sub_src_root, sub_distdir)
-        self.run_dist_scripts()
-        output_names = []
-        for a in archives:
-            compressed_name = self.distdir + archive_extension[a]
-            shutil.make_archive(self.distdir, a, root_dir=self.dist_sub, base_dir=self.dist_name)
-            output_names.append(compressed_name)
-        windows_proof_rmtree(self.distdir)
-        return output_names
+    def is_repo(self, src_root: str) -> bool:
+        return is_git(src_root)
 
 
 class HgDist(Dist):
+
+    def _check_output(self, cmd: T.Union[list, str]) -> str:
+        if isinstance(cmd, str):
+            cmd = cmd.split()
+        if cmd[0] != 'hg':
+            cmd.insert(0, 'hg')
+        return subprocess.check_output(cmd, cwd=self.src_root, text=True, encoding='utf-8')
```

Comment:
This helper method looks like a very bad idea. Passing a string and tokenizing it via a simple `.split()` is wrong. All this seems to do, is to prepend `hg` to the command line if it is not present, but I don't see why it cannot be simply specified at the call sites.

Reply:
I tried to answer to your comment with three changes:
- use `shlex.split`
- move this helper method in `Dist` and make it more general (i.e. potentially usable in `GitDist`)
- Add `"hg"` at the call sites

I hope it's going to be more acceptable.

I didn't remove this method because I don't understand why it would be such a very bad idea to have it. I understand the potential problem with `.split` but it seems to me that its presence and usage increase the overall quality of the code in `HgDist`.

Reply:
> I tried to answer to your comment with three changes:
> 
> * use `shlex.split`

This is still wrong.

> I didn't remove this method because I don't understand why it would be such a very bad idea to have it.

What does happen if one of the arguments you pass to `hg` has a space in it? As you wrote it, this method is not general, but works only for the command strings for which you use it, which you know do not have parameters that may contain spaces. However, having an helper method with a generic name work only is a very limited number of cases is not good.

> I understand the potential problem with `.split` but it seems to me that its presence and usage increase the overall quality of the code in `HgDist`.

Why do you thing that writing
```python
self._check_output('hg status meson.build --all --template {path}')
```
is better than writing
```python
subprocess.check_output(['hg', 'status', 'meson.build', '--all', '--template', '{path}'], cwd=self.src_root, universal_newlines=True)
```
?

Please note that the on;ly sensible way to pass the command to is via a list of strings, thus the only proper way to write the first is actually
```python
self._check_output(['hg', 'status', 'meson.build', '--all', '--template', '{path}'])
```
At that point, the only thing that the helper method does, it to hide the `cwd` and `universal_newlines` arguments, at the cost of requiring to look up the implementation of the helper method to understand what the code does.

Also note that your use of `encoding='utf-8'` is most likely wrong, unless Mercurial does not force UTF8 encoding also on systems that use different system encoding, which I think it is not the case.

Reply:
> What does happen if one of the arguments you pass to hg has a space in it?

I don't understand your point. It is the purpose of the function `shlex.split`

```
In [2]: shlex.split('hg log -T "{node} {date}\n"')
Out[2]: ['hg', 'log', '-T', '{node} {date}\n']
```

So it's going to work.

Regarding your second question, yes I find the short version without `[` and `'` much more readable but I guess it is a matter of taste.

Regarding `encoding`, you might be right.

Anyway, I removed this helper function in [46316d4](https://github.com/mesonbuild/meson/pull/12840/commits/46316d45b9c119dfcb0191e87ed3cadb87cb91db).

Reply:
I had to readd `encoding='utf-8'` to avoid `EncodingWarning: 'encoding' argument not specified.`

I checked and  `text=True, encoding='utf-8'` is used in other files in Meson code.

I wonder why the warning is not triggered by the call in the Git implementation with `universal_newlines` which should just be an alias for `text` (https://github.com/mesonbuild/meson/blob/master/mesonbuild/mdist.py#L183).